### DISCUSSION TITLE
Tweaking the CSS of Mermaid.js flowcharts

### DISCUSSION THREAD

**Discussion by PaulRBerg** [facebook/docusaurus#8806] [Q&A]
I spent a good few hours trying to modify the global CSS that gets applied to the Mermaid diagrams on my Docusaurus website, to no avoid.

Here's what I want do:

- Change the default font size of all text labels in all diagrams
- Change the default radius of all [circles nodes](https://mermaid.js.org/syntax/flowchart.html#a-node-in-the-form-of-a-circle)

I added the following CSS in `src/css/custom.css`:

```css
.mermaid .node circle {
  r: 100 !important;
}
```

I tried to provide a [`themeVariables`](https://mermaid.js.org/config/theming.html#theme-variables) object, like this:

```js
mermaid: {
  theme: { light: "neutral", dark: "dark" },
  themeVariables { ... }
},
```

That failed with this error:

> [ERROR] ValidationError: "mermaid.themeVariables" is not allowed

I tried specifying the `fontSize`, like this:

```js
mermaid: {
  options: {
    fontSize: 5,
  },
  theme: { light: "neutral", dark: "dark" },
},
```

That has no effect on the rendered diagrams (note: I have tried setting the theme to "base", as per their [indications](https://mermaid.js.org/config/theming.html#available-themes) in the docs).

What am I missing? Is it not possible to do what I want with the current Docusaurus integration of Mermaid.js?

Here's an example that works for a particular diagram:

````md
```mermaid
flowchart LR;
    foo((Foo));
    style foo r:40;
```
````

But having to repeat this styling for every single note in every single diagram would be quite a chore.

**Comment by slorber** [comment]
Hey

I'm not a mermaid expert althoughI built the Docusaurus integration.

As far as I understand, you can modify mermaid diagrams by using init directives.

https://docusaurus.io/tests/pages/diagrams#sequence-diagram-forest-theme-directive

You can also swizzle the Mermaid renderer component and use any Mermaid option here:

![CleanShot 2023-03-23 at 11 48 03@2x](https://user-images.githubusercontent.com/749374/227180588-62a3fad9-3f6f-4b78-9ca4-eb6fac0916df.png)

I am able to change the `fontSize` this way.

To be honest I'm not 100% sure the way we merge default theme/options in user-provided options is correct, this would require investigating a bit more but maybe the spread is not in the correct place here? 🤷‍♂️ 

```js
export function useMermaidConfig(): MermaidConfig {
  const {colorMode} = useColorMode();
  const mermaidThemeConfig = useMermaidThemeConfig();

  const theme = mermaidThemeConfig.theme[colorMode];
  const {options} = mermaidThemeConfig;

  return useMemo(
    () => ({startOnLoad: false, ...options, theme}), // Bad merge ???
    [theme, options],
  );
}

export function useMermaidSvg(
  txt: string,
  mermaidConfigParam?: MermaidConfig,
): string {

  const defaultMermaidConfig = useMermaidConfig();
  const mermaidConfig = mermaidConfigParam ?? defaultMermaidConfig; // no merge ???

  // ...
}
```

Try to change this locally maybe and tell us what works best?

**Reply by PaulRBerg** [reply]
Thank you for the thoughtful answer, @slorber!

I spent a few hours investigating this, and at this point, I am pretty sure that the problem is Mermaid itself, not Docusaurus. The issue is that `fontSize` does not work for flowcharts: https://github.com/mermaid-js/mermaid/issues/2896.

Your merging logic and use of the spread operator are correct:

<img width="521" alt="Screenshot 2023-03-24 at 5 04 34 PM" src="https://user-images.githubusercontent.com/8782666/227571278-fc6f545a-adca-481d-b3c2-dbe645630af1.png">

What would be worth doing in Docusaurus is to update the link [here](https://docusaurus.io/docs/next/markdown-features/diagrams#configuration) to point to [this URL](https://github.com/mermaid-js/mermaid/blob/e4d2118d4bfa023628a020b7ab1f8c491e6dc523/packages/mermaid/src/config.type.ts#L36) instead of [this URL](https://mermaid.js.org/config/configuration.html).

TBH, the Mermaid.js configuration page has left me more confused than I was before reading it. What matters for Docusaurus users is to find out *which config options* can be passed to `themeConfig.mermaid.options`; the `MermaidConfig` type provides this information.

**Reply by slorber** [reply]
Please send a PR with your suggested docs change. I'll be on holiday but will review it soon.

Linking to both may also work?

**Reply by PaulRBerg** [reply]
> Please send a PR

Will do!

> Linking to both may also work?

Yeah, I think the key question is what git ref to use in the URL. Perhaps the version tag?

**Reply by PaulRBerg** [reply]
Ended up using the version tag in https://github.com/facebook/docusaurus/pull/8830.
